lib/orderbook/MatchingEngine.ts

```diff
@@ -66,34 +101,140 @@ class MatchingEngine {
-    const addTo = isBuyOrder ? this.buyOrders : this.sellOrders;
-
     const matchingResult = this.match(order);
     if (matchingResult.remainingOrder && !discardRemaining) {
-      addTo.add(matchingResult.remainingOrder);
+      this.addOwnOrder(matchingResult.remainingOrder);
     }
 
     return matchingResult;
   }
 
+  private addOwnOrder = (order: StampedOwnOrder): boolean => {
+    return this.addOrder(order, this.ownOrders);
+  }
+
+  private addOrder = (order: StampedOrder, lists: OrderSidesLists<StampedOrder>): boolean => {
+    const isBuyOrder = order.quantity > 0;
+    const list = isBuyOrder ? lists.buy : lists.sell;
+    const queue = isBuyOrder ? this.queues.buy : this.queues.sell;
+
+    if (list.has(order.id)) {
+      return false;
+    } else {
+      list.set(order.id, order);
+      queue.add(order);
+      return true;
+    }
+  }
+
+  public removePeerOrderQuantity = (orderId: string, quantityToDecrease?: number): StampedPeerOrder | undefined => {
+    const order = this.peerOrders.buy.get(orderId) || this.peerOrders.sell.get(orderId);
+    if (!order) {
+      return;
+    }
+
+    if (quantityToDecrease && quantityToDecrease < Math.abs(order.quantity)) {
+      order.quantity = order.quantity - quantityToDecrease;
+      return { ...order, quantity: quantityToDecrease };
+    } else {
+      this.removePeerOrder(order);
+      return order;
+    }
+  }
+
+  /**
+   * Remove all orders given a peer pubKey.
+   */
+  public removePeerOrders = (peerPubKey: string): StampedPeerOrder[] => {
+    const callback = (order: StampedOrder) => (order as StampedPeerOrder).peerPubKey === peerPubKey;
+
+    // remove from queues
+    const removedOrders = [
+      ...this.queues.buy.removeMany(callback),
+      ...this.queues.sell.removeMany(callback),
+    ] as StampedPeerOrder[];
+
+    // remove from lists. for further optimization, we can maintain a separate list for each peer pubKey
+    removedOrders.forEach((order: StampedPeerOrder) => {
+      const list = order.quantity > 0 ? this.peerOrders.buy : this.peerOrders.sell;
+      list.delete(order.id);
+    });
+
+    return removedOrders;
+  }
+
+  public removeOwnOrder = (orderId: string): StampedOwnOrder | undefined => {
+    const order = this.ownOrders.buy.get(orderId) || this.ownOrders.sell.get(orderId);
+    if (!order) {
+      return;
+    }
+
+    this.removeOrder(order, this.ownOrders);
+    return order;
+  }
+
+  private removePeerOrder = (order: StampedPeerOrder) => {
+    this.removeOrder(order, this.peerOrders);
+  }
+
+  private removeOrder = (order: StampedOrder, lists: OrderSidesLists<StampedOrder>) => {
+    const isBuyOrder = order.quantity > 0;
+    const list = isBuyOrder ? lists.buy : lists.sell;
+    const queue = isBuyOrder ? this.queues.buy : this.queues.sell;
+
+    list.delete(order.id);
+    queue.remove(order);
+  }
+
+  private getOrderList = (order: StampedOrder): OrderList<StampedOrder> => {
+    switch (isOwnOrder(order)) {
```

Comment:
A `switch` is kinda overkill for a boolean isn't it?

Reply:
Indeed, I don't know why I thought it would be cleaner like this. Will change it.